### DISCUSSION TITLE
Webhook returns nothing

### DISCUSSION THREAD

**Discussion by JekaSylar** [westacks/telebot#71] [Q&A]
Good afternoon, for a week now I can’t figure out why the webhook doesn’t return anything to me?
Tried on a live server, tried through polling
Here is the code
web.php
```php
Route::post('/telebot/webhook/bot/{token}', [TelegramBotAdmarController::class, 'webhook']);
``` 
telebot.php
```php

    'bots' => [
        'admar' => [
            'token' => env('TELEGRAM_BOT_TOKEN'),
            'name' => env('TELEGRAM_BOT_NAME', null),
            'api_url' => env('TELEGRAM_API_URL', 'https://api.telegram.org/bot{TOKEN}/{METHOD}'),
            'exceptions' => true,
            'async' => false,

            'webhook' => [
                'url' => env(
                    'TELEGRAM_BOT_WEBHOOK_URL',
                    env('APP_URL') . '/telebot/webhook/bot/' . env('TELEGRAM_BOT_TOKEN')
                ),
                // 'certificate'       => env('TELEGRAM_BOT_CERT_PATH', storage_path('app/ssl/public.pem')),
                // 'ip_address'        => '8.8.8.8',
                // 'max_connections'   => 40,
                // 'allowed_updates'   => ["message", "edited_channel_post", "callback_query"],
                // 'secret_token'      => env('TELEGRAM_KEY', null),
            ],

            'poll' => [
                'limit' => 100,
                'timeout' => 0,
                'allowed_updates' => ["message", "edited_channel_post", "callback_query"]
            ],

            'handlers' => [
                // Your update handlers
            ],
        ],
``` 
TelegramBotAdmarController.php
```php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;
use WeStacks\TeleBot\Laravel\TeleBot;

use function MongoDB\BSON\toJSON;

class TelegramBotAdmarController extends Controller
{
    public function webhook($token)
    {
        $test = TeleBot::bot('admar')->handleUpdate();
        Log::info('message', (array)$test);
    }


    public function updateMessage()
    {
        $handler = function (TeleBot $bot, Update $update, $next) {
            if ($update->message->text === '/start') {
                return $bot->sendMessage([
                    'chat_id' => $update->chat()->id,
                    'text' => 'Hello, World!'
                ]);
            }

            return $next();
        };

        Log::info('telegbot - ' . toJSON($handler));
    }


}

``` 

but for some reason I can't write anything to the log

**Comment by punyflash** [comment]
Please read documentation more precise, you have not registered any handler: https://westacks.github.io/telebot/#/updates
```php
'handlers' => [
        // Your update handlers
],
```

**Reply by JekaSylar** [reply]
I wrote like this
```php
<?php

namespace App\Http\Controllers;


use Illuminate\Http\Request;
use Illuminate\Support\Facades\Log;
use WeStacks\TeleBot\Laravel\Facades\TeleBot;
use WeStacks\TeleBot\Objects\Update;


class TelegramBotAdmarController extends Controller
{
    public function webhook(Request $request, $token)
    {
        $bot = TeleBot::bot('admar');
        $bot->setWebhook(['url' => $request->url()]);

        $updates = $bot->handleWebhook($request->getContent());
        foreach ($updates as $update) {
            $this->handleUpdate($bot, $update);
        }

        return response('Webhook processed', 200);
    }


    public function handleUpdate(TeleBot $bot, Update $update)
    {
        if ($update->message->text === '/start') {
            Log::info('telegbot start - ' . $update->message->text);

            return $bot->sendMessage([
                'chat_id' => $update->chat()->id,
                'text' => 'Hello, World!',
            ]);
        } else {
            Log::info('telegbot - ' . $update->message->text);
        }
    }


}
```
uploaded to server 
I execute command 
```
php artisan telebot:webhook                   
No task specified!
```
And I get an error

**Reply by punyflash** [reply]
What you are doing makes no sense. First of all you don't need controller for your bot, library has it within already - just use handler classes and register them in a `config/telebot.php`

**Reply by JekaSylar** [reply]
Я понял что вы имеете введу я создал контроле в нем метод а как мне в handlers передать его 
'handlers' => [
                \App\Http\Controllers\TelegramBotAdmarController::class,
                
            ],

**Reply by JekaSylar** [reply]
> То, что вы делаете, не имеет смысла. Во-первых, вам не нужен контроллер для вашего бота, он уже есть в библиотеке — просто используйте классы-обработчики и зарегистрируйте их в`config/telebot.php`

Please tell me why when I transfer the code to the web server I use the webhook, it does not work
```php
class TelegramBotAdmarController extends Controller
{
    public function webhook(Request $request, $token)
    {
        $bot = TeleBot::bot('admar');
        $bot->setWebhook(['url' => $request->url()]);
    }


}
```
although locally everything through polling works

**Reply by punyflash** [reply]
Again - you don't need to create any controller or webhook handler for your bot. Just set your `APP_URL` in `.env`, run `php artisan telebot:webhook --setup` via SSH on your server and everything shoud work.

**Reply by JekaSylar** [reply]
1. that is, remove the route that I created?
2. in the settings APP_URL to register the site address?
3. execute the command and will it work?
Right ?

**Reply by punyflash** [reply]
Yes, library have everything within already:
https://github.com/westacks/telebot/blob/c00ad32a35e0b5064ac343b08db601865852e283/src/Laravel/routes/telebot.php#L15-L16
https://github.com/westacks/telebot/blob/c00ad32a35e0b5064ac343b08db601865852e283/src/Laravel/Controllers/WebhookController.php#L9-L15
https://github.com/westacks/telebot/blob/c00ad32a35e0b5064ac343b08db601865852e283/src/Laravel/Artisan/WebhookCommand.php#L46-L74

**Reply by JekaSylar** [reply]
I did as you wrote but it still does not return anything ((

**Reply by JekaSylar** [reply]
| | |
|---|---|
| Url | https://admar.com.ua/telebot/webhook/bot/623:8_O0vHg |
| Has Custom Certificate | No |                                        
| Pending Update Count | 5 |
| Last Error Date  | 1685389375 |
| Last Error Message | Wrong response from the webhook: 403 Forbidden |
| Max Connections | 40 |
| Ip Address | 185.166.217.119 |

**Reply by punyflash** [reply]
Your webhook returns 403. Check your configuration. Bot token does not seem right from your URL. And don't send it here if you don't want it compromised.

**Comment by JekaSylar** [comment]
In which method to write the handler?

**Comment by JekaSylar** [comment]
```php
 'bots' => [
        'admar' => [
            'token' => env('TELEGRAM_BOT_TOKEN'),
            'name' => env('TELEGRAM_BOT_NAME', null),
            'api_url' => env('TELEGRAM_API_URL', 'https://api.telegram.org/bot{TOKEN}/{METHOD}'),
            'exceptions' => true,
            'async' => true,

            'webapp' => [
                'redirect_url' => '/auth/telegram/callback',
            ],

            'webhook' => [
                'url' => env(
                    'TELEGRAM_BOT_WEBHOOK_URL',
                    env('APP_URL') . '/api/telebot/webhook/bot/' . env('TELEGRAM_BOT_TOKEN')
                ),
                // 'certificate'       => env('TELEGRAM_BOT_CERT_PATH', storage_path('app/ssl/public.pem')),
                // 'ip_address'        => '8.8.8.8',
                // 'max_connections'   => 40,
                // 'allowed_updates'   => ["message", "edited_channel_post", "callback_query"],
                // 'secret_token'      => env('TELEGRAM_KEY', null),
            ],

            'poll' => [
                'limit' => 100,
                'timeout' => 0,
                'allowed_updates' => ["message", "edited_channel_post", "callback_query"]
            ],

            'handlers' => [
                // \App\Telegram\Handlers\AdmarHandler::class,
                \App\Telegram\Handlers\Admar\StatAdmarBotHandler::class,


            ],
        ],
```

**Comment by mohamedhekal** [comment]
It seems like you're trying to set up a Telegram bot webhook using Laravel and the TeleBot package, but you're not receiving any updates in your webhook. There could be several reasons for this issue. Let's go through some troubleshooting steps to help you identify and resolve the problem:

1. Verify Webhook URL:
   Ensure that the webhook URL you have configured in your bot's settings on Telegram matches the one you've set in your Laravel configuration. Double-check that there are no typos or trailing slashes.

2. SSL Certificate:
   If you're using an HTTPS webhook URL, make sure you have a valid SSL certificate installed on your server. Telegram requires a secure connection (HTTPS) for webhooks.

3. Token Configuration:
   Verify that the `TELEGRAM_BOT_TOKEN` environment variable in your `.env` file is correctly set and matches the token for your Telegram bot.

4. Routing Configuration:
   Ensure that your route definition in `web.php` is correctly set up to handle POST requests to the webhook URL. Make sure that the `{token}` parameter is correctly passed.

5. Logging Configuration:
   Check your Laravel logging configuration (`config/logging.php`) to ensure that the log channel you're using is correctly configured, and the log file has the necessary permissions to write to. You can also check Laravel's log files in the `storage/logs` directory for any error messages.

6. Debugging:
   To debug the issue further, you can add some debug statements to your `TelegramBotAdmarController.php` file. For example, you can add the following lines to your `webhook` method to log the incoming request data:

   ```php
   public function webhook($token)
   {
       Log::info('Webhook received');
       Log::info('Request data: ' . json_encode(request()->all()));
       // The rest of your code
   }
   ```

   This will help you see if your webhook is receiving requests from Telegram.

7. ngrok (Optional):
   You can use a tool like ngrok to temporarily expose your local server to the internet and test the webhook locally. This can help you verify that your local setup is working correctly before deploying it to a live server.

8. Error Handling:
   Implement error handling in your webhook code to catch any exceptions or errors that might be preventing the webhook from processing updates. Log any errors for further analysis.

By following these steps and adding debugging statements, you should be able to identify the issue and get your Telegram bot webhook working correctly. Make sure to monitor your server logs for any errors or issues that may arise during the webhook setup process.
